doc/designs/ech-api.md

```diff
@@ -0,0 +1,590 @@
+returned. If no ECHConfig found has supported options then none will be
+returned and the function will return 0.
+
+After a call to ``OSSL_ech_find_echconfigs()``, the application can make a
+sequence of calls to ``SSL_ech_set1_echconfig()`` for each of the ECHConfig
+values found.  (The various output buffers must be freed by the client
+afterwards, see the example code in
+[``test/ech_test.c``](https://github.com/sftcd/openssl/blob/ECH-draft-13c/test/ech_test.c).)
+
+Clients can additionally more directly control the values to be used for inner
+and outer SNI and ALPN values via specific APIs. This allows a client to
+override the ``public_name`` present in an ECHConfigList that will otherwise
+be used for the outer SNI. The ``no_outer`` input allows a client to emit an
+outer CH with no SNI at all.
+
+```c
+int SSL_ech_set_server_names(SSL *s, const char *inner_name,
+                             const char *outer_name, int no_outer);
+int SSL_ech_set_outer_server_name(SSL *s, const char *outer_name, int no_outer);
+int SSL_ech_set_outer_alpn_protos(SSL *s, const unsigned char *protos,
+                                  unsigned int protos_len);
+int SSL_CTX_ech_set_outer_alpn_protos(SSL_CTX *s, const unsigned char *protos,
+                                      unsigned int protos_len);
+```
+
+If a client attempts ECH but that fails, or sends an ECH-GREASEd CH, to
+an ECH-supporting server, then that server may return an ECH "retry-config"
+value that the client could choose to use in a subsequent connection. The
+client can detect this situation via the ``SSL_ech_get_status()`` API and
+can access the retry config value via:
+
+```c
+int SSL_ech_get_retry_config(SSL *s, unsigned char **ec, size_t *eclen);
+```
+
+Clients that need fine control over which ECHConfig (from those available) will
+be used, can query the SSL connection, retrieving information about the set of
+ECHConfig values available, and then, if desired, down-select to one of those,
+e.g., based on the ``public_name`` that will be used. This would enable a
+client that selects the server address to use based on IP address hints that
+can also be present in an HTTPS/SCVB resource record to ensure that the correct
+matching ECHConfig is used. The information is presented to the client using
+the ``OSSL_ECH_INFO`` type, which provides a simplified view of ECHConfig data,
+but where each element of an array corresponds to exactly one ECH public value
+and set of names.
+
+```c
+/*
+ * Application-visible form of ECH information from the DNS, from config
+ * files, or from earlier API calls. APIs produce/process an array of these.
+ */
+typedef struct ossl_ech_info_st {
+    int index; /* externally re-usable reference to this value */
+    char *public_name; /* public_name from API or ECHConfig */
+    char *inner_name; /* server-name (for inner CH if doing ECH) */
+    char *outer_alpns; /* outer ALPN string */
+    char *inner_alpns; /* inner ALPN string */
```

Comment:
It seems like ALPN is strictly speaking binary data (although all real ALPN values are printable strings AFAIK) - but in theory at least don't we need a length value for these?

Reply:
In theory, yes, and indeed most other APIs handling alpns also pass a length though e.g. `SSL_SESSION_get0_alpn_selected()`  doesn't seem to. I've added a length as that seems better all right.